app/scripts/lib/multichain-api/multichainSubscriptionManager.ts

```diff
@@ -0,0 +1,153 @@
+import {
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+} from '@metamask/network-controller';
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { parseCaipChainId } from '@metamask/utils';
+import { toHex } from '@metamask/controller-utils';
+import { Scope } from './scope';
+
+type SubscriptionManager = {
+  events: {
+    on: (
+      event: string,
+      listener: MultichainSubscriptionManager['onNotification'],
+    ) => void;
+    off: (
+      event: string,
+      listener: MultichainSubscriptionManager['onNotification'],
+    ) => void;
+  };
+  destroy?: () => void;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
+const createSubscriptionManager = require('@metamask/eth-json-rpc-filters/subscriptionManager');
+
+type MultichainSubscriptionManagerOptions = {
+  findNetworkClientIdByChainId: NetworkControllerFindNetworkClientIdByChainIdAction['handler'];
+  getNetworkClientById: NetworkControllerGetNetworkClientByIdAction['handler'];
+};
+
+export default class MultichainSubscriptionManager extends SafeEventEmitter {
+  private subscriptionsByChain: {
+    [scope: string]: {
+      [domain: string]: MultichainSubscriptionManager['onNotification'];
```

Comment:
technically this type is incorrect as we're assigning a instance of the function where the params are fully bound